### PR TITLE
[SYSTEMML-1931] Support logical operators XOR over matrices, and scalars

### PR DESCRIPTION
**Pull request by j143**
_what is here?_
**Supports:**
operations over scalars, all combinations for `AND`, `XOR`, `NOT`, `XOR` are working, such as 
``` 
say A = matrix("1 1
                0 0", rows=2, cols=2);
      n = 1.0;
      m = TRUE;
     1. xor(m,n) gives 0.0
     2. xor( m, !n) gives TRUE
     3. ( m & !n) gives FALSE
     4. ( m | !n) gives TRUE
     5. xor(A, A) gives all zeros

     6. xor(A,m) or xor(A, n) - not working
     7. (A & A), (A | A), (!A) - not working
```

**what I've tried to change?**
1. changed function objects, to support `(double, double)` to return `double` -- all operators
2. Tried to handle `BuiltinBinary` as `BooleanBinary` operations, and added separted classes `MatrixMatrixBooleanCPInstruction`, `MatrixScalarBooleanCPInstruction`, `ScalaScalarBooleanCPInstruction`. 
 
3. Handled the `!` case (`BuiltinUnary`), as `BooleanUnaryCPInstruction`, for the matrix case.
4. At the end added a `junit` test, `ElementwiseXorTest`
   - I've done some mistake instruction gen for spark
   - runtime error in mr case, for all xor over matrices

### CODE REVIEW

File: src/main/java/org/apache/sysml/runtime/instructions/cp/MatrixMatrixBooleanCPInstruction.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+package org.apache.sysml.runtime.instructions.cp;
+
+import org.apache.sysml.runtime.DMLRuntimeException;
+import org.apache.sysml.runtime.controlprogram.context.ExecutionContext;
+import org.apache.sysml.runtime.matrix.data.MatrixBlock;
+import org.apache.sysml.runtime.matrix.operators.BinaryOperator;
+import org.apache.sysml.runtime.matrix.operators.Operator;
+
+public class MatrixMatrixBooleanCPInstruction extends BooleanBinaryCPInstruction {
```
Comment (j143) [inline src/main/java/org/apache/sysml/runtime/instructions/cp/MatrixMatrixBooleanCPInstruction.java]: Can this class be kept only to handle logical operators, as total of 10 operators may justify this.

**Comment by j143** [comment]
Hi @mboehm7 , If possible can the following changes made to pr & merged after 
  1. the adjustment for xor operator
  2. help me find why matrix, matrix are not working for other operators

My mind is all with [SYSTEMML-979], I want to complete it by `7th Jan`.

Thanks,

**Comment by akchinSTC** [comment]
Build failed, see build log for details

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2226/<h2>Failed Tests: <span class='status-failure'>18</span></h2><h3><a name='SystemML-PullRequestBuilder/org.apache.systemml:systemml' /><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2226/org.apache.systemml$systemml/testReport'>SystemML-PullRequestBuilder/org.apache.systemml:systemml</a>: <span class='status-failure'>18</span></h3><ul><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2226/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseXorTest/testLargeXorSparseHybrid/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseXorTest.testLargeXorSparseHybrid</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2226/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseXorTest/testXorDenseMR/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseXorTest.testXorDenseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2226/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseXorTest/testXorDenseSP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseXorTest.testXorDenseSP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2226/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseXorTest/testLargeXorDenseMR/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseXorTest.testLargeXorDenseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2226/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseXorTest/testLargeXorDenseSP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseXorTest.testLargeXorDenseSP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2226/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseXorTest/testXorSparseMR/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseXorTest.testXorSparseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2226/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseXorTest/testXorSparseSP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseXorTest.testXorSparseSP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2226/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseXorTest/testLargeXorDenseHybrid/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseXorTest.testLargeXorDenseHybrid</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2226/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseXorTest/testLargeXorSparseMR/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseXorTest.testLargeXorSparseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2226/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseXorTest/testLargeXorSparseSP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseXorTest.testLargeXorSparseSP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2226/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.reorg/FullReverseTest/testReverseVectorSparseMR/'><strong>org.apache.sysml.test.integration.functions.reorg.FullReverseTest.testReverseVectorSparseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2226/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.reorg/FullReverseTest/testReverseVectorSparseSP/'><strong>org.apache.sysml.test.integration.functions.reorg.FullReverseTest.testReverseVectorSparseSP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2226/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.reorg/FullReverseTest/testReverseVectorDenseMR/'><strong>org.apache.sysml.test.integration.functions.reorg.FullReverseTest.testReverseVectorDenseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2226/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.reorg/FullReverseTest/testReverseVectorDenseSP/'><strong>org.apache.sysml.test.integration.functions.reorg.FullReverseTest.testReverseVectorDenseSP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2226/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.reorg/FullReverseTest/testReverseMatrixSparseMR/'><strong>org.apache.sysml.test.integration.functions.reorg.FullReverseTest.testReverseMatrixSparseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2226/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.reorg/FullReverseTest/testReverseMatrixSparseSP/'><strong>org.apache.sysml.test.integration.functions.reorg.FullReverseTest.testReverseMatrixSparseSP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2226/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.reorg/FullReverseTest/testReverseMatrixDenseMR/'><strong>org.apache.sysml.test.integration.functions.reorg.FullReverseTest.testReverseMatrixDenseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2226/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.reorg/FullReverseTest/testReverseMatrixDenseSP/'><strong>org.apache.sysml.test.integration.functions.reorg.FullReverseTest.testReverseMatrixDenseSP</strong></a></li></ul>

**Comment by mboehm7** [comment]
sure I'll finish it up and then merge it later today.

**Comment by mboehm7** [comment]
ok just to explain what was missing - here is what I changed:
* rewrote the test to compare against R and test AND, OR, XOR, NOT (the expected matrix computation was actually wrong)
* fixed the runtime operations (the implementations did not correctly convert doubles to boolean and implemented OR as AND)
* fixed missing compiler integration for AND, OR, and NOT (compiled always scalar outputs)
* fixed missing runtime integration into MR and SPARK instructions
* removed new instructions and redirected AND, OR, and XOR to arithmetic instructions which otherwise would be an exact duplicate.

The integration of NOT is a bit trickier and will follow tomorrow.

**Comment by mboehm7** [comment]
thanks again for the initial version @j143 - I'll merge

**Comment by j143** [comment]
Hi @mboehm7 , I am sorry for my mistakes, and will try to improve in the next pr. Thanks a lot for the correcting & guiding. 

This will help me as a template for the next bitwise operations, which I will submit tomorrow(approx. 20 hrs from this message).

Thanks.

**Pull request closed without merging**